File: backend/app/services/admin_service.py
"""Super-admin APIs — privacy-limited farmer summaries (no soil chemistry)."""

from __future__ import annotations

import re
import uuid

from sqlalchemy import desc, func, or_, select
from sqlalchemy.orm import Session

from app.core.config import get_settings
from app.core.exceptions import ForbiddenError, NotFoundError
from app.db.models import FarmProfile, RecommendationRun, SoilReading, UserAccount, UserRole
from app.services.district_utils import resolve_district

# Automated test / seed accounts — never shown in live admin mode
_TEST_EMAIL = re.compile(
    r"(@example\.com$|@farmsense\.test$|^(test-|reset-|e2e-|exists-|new-))",
    re.IGNORECASE,
)
_TEST_DISTRICTS = {"test district", "reset district", "reset region", "test region"}


def _is_demo_farmer(user: UserAccount, farm: FarmProfile | None = None) -> bool:
    email = (user.email or "").strip().lower()
    if _TEST_EMAIL.search(email):
        return True
    if farm:
        for label in (farm.district_name, farm.region_label):
            if label and label.strip().lower() in _TEST_DISTRICTS:
                return True
    return False


def _live_farmer_filter():
    return UserAccount.role == UserRole.farmer


def _district_for_farm(farm: FarmProfile | None) -> str:
    if not farm:
        return "Unknown"
    raw = farm.district_name or farm.region_label
    district = resolve_district(raw, use_api=True)
    # Persist a cleaner short name when the stored value was a full address
    if farm.district_name and district and farm.district_name.strip() != district:
        if "," in farm.district_name or len(farm.district_name) > len(district) + 8:
            farm.district_name = district
            if farm.region_label and (
                "," in farm.region_label or "United Kingdom" in farm.region_label
            ):
                farm.region_label = district
    return district


def _plan_flags(run: RecommendationRun | None) -> dict:
    if not run:
        return {
            "hasRecommendations": False,
            "planStatus": "none",
            "finalized": False,
            "topCrop": None,
            "lastPlanAt": None,
        }
    ranked = run.ranked_output or {}
    finalized = bool(ranked.get("finalized")) or run.status.value == "completed"
    plan_status = ranked.get("planStatus") or ("finalized" if finalized else "draft")
    return {
        "hasRecommendations": True,
        "planStatus": plan_status,
        "finalized": finalized,
        "topCrop": run.top_crop,
        "lastPlanAt": run.created_at.isoformat() if run.created_at else None,
    }


def _farmer_summary(db: Session, user: UserAccount, farm: FarmProfile | None) -> dict:
    has_soil = False
    soil_count = 0
    latest_run = None
    district = _district_for_farm(farm)
    if farm:
        soil_count = (
            db.scalar(
                select(func.count())
                .select_from(SoilReading)
                .where(SoilReading.farm_profile_id == farm.id)
            )
            or 0
        )
        has_soil = soil_count > 0
        latest_run = db.scalar(
            select(RecommendationRun)
            .where(RecommendationRun.farm_profile_id == farm.id)
            .order_by(desc(RecommendationRun.created_at))
            .limit(1)
        )

    plan = _plan_flags(latest_run)
    return {
        "id": str(user.id),
        "name": user.full_name,
        "email": user.email,
        "isActive": user.is_active,
        "createdAt": user.created_at.isoformat() if user.created_at else None,
        "lastLoginAt": user.last_login_at.isoformat() if user.last_login_at else None,
        "district": district,
        "region": district,
        "farmSize": float(farm.area_hectares) if farm and farm.area_hectares else None,
        "countryCode": farm.country_code if farm else None,
        "hasSoilData": has_soil,
        "soilReadingsCount": soil_count,
        **plan,
    }


def admin_overview(db: Session) -> dict:
    users = db.scalars(select(UserAccount).where(_live_farmer_filter())).all()
    real_users: list[tuple[UserAccount, FarmProfile | None]] = []
    for user in users:
        farm = db.scalar(select(FarmProfile).where(FarmProfile.user_id == user.id).limit(1))
        if _is_demo_farmer(user, farm):
            continue
        real_users.append((user, farm))

    total = len(real_users)
    active = sum(1 for u, _ in real_users if u.is_active)
    with_soil = 0
    draft_plans = 0
    finalized_plans = 0
    district_counts: dict[str, int] = {}

    for user, farm in real_users:
        district = _district_for_farm(farm)
        district_counts[district] = district_counts.get(district, 0) + 1

        if not farm:
            continue

        soil_n = (
            db.scalar(
                select(func.count())
                .select_from(SoilReading)
                .where(SoilReading.farm_profile_id == farm.id)
            )
            or 0
        )
        if soil_n > 0:
            with_soil += 1

        latest = db.scalar(
            select(RecommendationRun)
            .where(RecommendationRun.farm_profile_id == farm.id)
            .order_by(desc(RecommendationRun.created_at))
            .limit(1)
        )
        flags = _plan_flags(latest)
        if flags["planStatus"] == "draft":
            draft_plans += 1
        elif flags["finalized"] or flags["planStatus"] == "finalized":
            finalized_plans += 1

    db.commit()

    by_district = [
        {"district": name, "farmers": count}
        for name, count in sorted(district_counts.items(), key=lambda x: (-x[1], x[0]))
    ]

    return {
        "mode": "live",
        "totals": {
            "farmers": total,
            "active": active,
            "inactive": total - active,
            "withSoilData": with_soil,
            "draftPlans": draft_plans,
            "finalizedPlans": finalized_plans,
        },
        "byDistrict": by_district,
    }


def list_farmers(db: Session, *, q: str | None = None, page: int = 1, limit: int = 20) -> dict:
    page = max(1, page)
    limit = min(100, max(1, limit))

    stmt = select(UserAccount).where(_live_farmer_filter())
    if q and q.strip():
        term = f"%{q.strip().lower()}%"
        stmt = stmt.where(
            or_(
                func.lower(UserAccount.full_name).like(term),
                func.lower(UserAccount.email).like(term),
            )
        )
    stmt = stmt.order_by(desc(UserAccount.created_at))
    users = db.scalars(stmt).all()

    summaries = []
    for user in users:
        farm = db.scalar(select(FarmProfile).where(FarmProfile.user_id == user.id).limit(1))
        if _is_demo_farmer(user, farm):
            continue
        summaries.append(_farmer_summary(db, user, farm))

    db.commit()

    total = len(summaries)
    offset = (page - 1) * limit
    items = summaries[offset : offset + limit]

    return {
        "mode": "live",
        "items": items,
        "page": page,
        "limit": limit,
        "total": total,
        "totalPages": max(1, (total + limit - 1) // limit) if total else 0,
    }


def get_farmer(db: Session, farmer_id: uuid.UUID) -> dict:
    user = db.get(UserAccount, farmer_id)
    if not user or user.role != UserRole.farmer:
        raise NotFoundError("Farmer not found")
    farm = db.scalar(select(FarmProfile).where(FarmProfile.user_id == user.id).limit(1))
    if _is_demo_farmer(user, farm):
        raise NotFoundError("Farmer not found")
    summary = _farmer_summary(db, user, farm)
    db.commit()
    return summary


def set_farmer_active(db: Session, farmer_id: uuid.UUID, is_active: bool) -> dict:
    settings = get_settings()
    user = db.get(UserAccount, farmer_id)
    if not user or user.role != UserRole.farmer:
        raise NotFoundError("Farmer not found")
    if user.email.strip().lower() == settings.super_admin_email.strip().lower():
        raise ForbiddenError("Cannot change the super-admin account from farmer controls")
    if _is_demo_farmer(user):
        raise NotFoundError("Farmer not found")

    user.is_active = bool(is_active)
    db.commit()
    db.refresh(user)
    farm = db.scalar(select(FarmProfile).where(FarmProfile.user_id == user.id).limit(1))
    return _farmer_summary(db, user, farm)


def delete_farmer(db: Session, farmer_id: uuid.UUID) -> dict:
    """Permanently delete a farmer account and cascaded farm data."""
    settings = get_settings()
    user = db.get(UserAccount, farmer_id)
    if not user or user.role != UserRole.farmer:
        raise NotFoundError("Farmer not found")
    if user.email.strip().lower() == settings.super_admin_email.strip().lower():
        raise ForbiddenError("Cannot delete the super-admin account")
    if user.role == UserRole.admin:
        raise ForbiddenError("Cannot delete an admin account")
    if _is_demo_farmer(user):
        raise NotFoundError("Farmer not found")

    email = user.email
    name = user.full_name
    db.delete(user)
    db.commit()
    return {"deleted": True, "id": str(farmer_id), "email": email, "name": name}


def _pct(part: int, whole: int) -> float:
    if whole <= 0:
        return 0.0
    return round((part / whole) * 100, 1)


def _build_insights(
    *,
    total: int,
    with_soil: int,
    finalized: int,
    draft: int,
    no_plan: int,
    inactive: int,
    by_district: list[dict],
    top_crops: list[dict],
) -> list[str]:
    if total == 0:
        return ["No farmers yet. Analytics will appear after the first registration."]

    insights: list[str] = []
    soil_pct = _pct(with_soil, total)
    fin_pct = _pct(finalized, total)

    if soil_pct >= 60:
        insights.append(f"Soil data adoption is strong at {soil_pct}% of farmers.")
    elif soil_pct > 0:
        insights.append(
            f"Only {soil_pct}% of farmers have submitted soil data — worth nudging new sign-ups to complete Plan."
        )
    else:
        insights.append("No soil submissions yet. Plan completion will unlock crop insights.")

    if fin_pct >= 40:
        insights.append(f"{fin_pct}% of farmers have a finalized crop plan.")
    elif draft > finalized:
        insights.append(
            f"More draft plans ({draft}) than finalized ({finalized}) — farmers may need help confirming picks."
        )
    elif no_plan == total:
        insights.append("No crop plans yet. Encourage farmers to start from Plan after registration.")

    if inactive:
        insights.append(f"{inactive} inactive account(s) are blocked from signing in.")

    if by_district:
        top = by_district[0]
        insights.append(f"Most farmers are in {top['district']} ({top['farmers']}).")

    if top_crops:
        lead = top_crops[0]
        insights.append(f"Leading finalized crop: {lead['crop']} ({lead['farmers']} farm(s)).")

    return insights[:5]


def admin_analytics(db: Session) -> dict:
    """Privacy-safe platform analytics for the super-admin Analysis page."""
    users = db.scalars(select(UserAccount).where(_live_farmer_filter())).all()
    rows: list[dict] = []
    for user in users:
        farm = db.scalar(select(FarmProfile).where(FarmProfile.user_id == user.id).limit(1))
        if _is_demo_farmer(user, farm):
            continue
        rows.append(_farmer_summary(db, user, farm))
    db.commit()

    total = len(rows)
    active = sum(1 for r in rows if r["isActive"])
    inactive = total - active
    with_soil = sum(1 for r in rows if r["hasSoilData"])
    finalized = sum(1 for r in rows if r.get("finalized") or r.get("planStatus") == "finalized")
    draft = sum(1 for r in rows if r.get("planStatus") == "draft" and not r.get("finalized"))
    no_plan = total - finalized - draft

    district_counts: dict[str, int] = {}
    crop_counts: dict[str, int] = {}
    month_counts: dict[str, int] = {}
    sizes: list[float] = []

    for r in rows:
        d = r.get("district") or "Unknown"
        district_counts[d] = district_counts.get(d, 0) + 1

        if r.get("finalized") and r.get("topCrop"):
            crop = str(r["topCrop"]).strip()
            if crop:
                crop_counts[crop] = crop_counts.get(crop, 0) + 1

        if r.get("createdAt"):
            month_key = r["createdAt"][:7]  # YYYY-MM
            month_counts[month_key] = month_counts.get(month_key, 0) + 1

        if r.get("farmSize") is not None:
            try:
                sizes.append(float(r["farmSize"]))
            except (TypeError, ValueError):
                pass

    by_district = [
        {"district": name, "farmers": count}
        for name, count in sorted(district_counts.items(), key=lambda x: (-x[1], x[0]))
    ]
    top_crops = [
        {"crop": name, "farmers": count}
        for name, count in sorted(crop_counts.items(), key=lambda x: (-x[1], x[0]))
    ][:8]

    signups = []
    for month in sorted(month_counts.keys()):
        year, mon = month.split("-")
        months = "Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec".split()
        label = f"{months[int(mon) - 1]} {year}"
        signups.append({"month": month, "label": label, "farmers": month_counts[month]})

    farm_size = {
        "avgHa": round(sum(sizes) / len(sizes), 2) if sizes else None,
        "minHa": round(min(sizes), 2) if sizes else None,
        "maxHa": round(max(sizes), 2) if sizes else None,
        "reported": len(sizes),
    }

    return {
        "mode": "live",
        "totals": {
            "farmers": total,
            "active": active,
            "inactive": inactive,
            "withSoilData": with_soil,
            "draftPlans": draft,
            "finalizedPlans": finalized,
            "noPlan": no_plan,
        },
        "rates": {
            "soilAdoptionPct": _pct(with_soil, total),
            "finalizedPct": _pct(finalized, total),
            "draftPct": _pct(draft, total),
            "noPlanPct": _pct(no_plan, total),
            "activePct": _pct(active, total),
        },
        "planStatus": [
            {"status": "Finalized", "count": finalized},
            {"status": "Draft", "count": draft},
            {"status": "No plan", "count": no_plan},
        ],
        "byDistrict": by_district,
        "topCrops": top_crops,
        "signupsByMonth": signups,
        "farmSize": farm_size,
        "insights": _build_insights(
            total=total,
            with_soil=with_soil,
            finalized=finalized,
            draft=draft,
            no_plan=no_plan,
            inactive=inactive,
            by_district=by_district,
            top_crops=top_crops,
        ),
    }

File: frontend/src/api/services/admin.service.js
import { backendEndpoints } from '../config'
import { backendClient } from '../backend/client'
import { withMockDelay } from '../mock/utils'

/** Sample-only farmers for Dummy mode (never mixed into Live). */
const mockFarmers = [
  {
    id: 'mock-f1',
    name: 'Alex Farmer',
    email: 'alex.demo@farmsense.local',
    isActive: true,
    createdAt: '2026-01-10T10:00:00Z',
    lastLoginAt: '2026-07-12T08:00:00Z',
    district: 'Northumberland',
    region: 'Northumberland',
    farmSize: 4.5,
    countryCode: 'GB',
    hasSoilData: true,
    soilReadingsCount: 2,
    hasRecommendations: true,
    planStatus: 'finalized',
    finalized: true,
    topCrop: 'Tomato',
    lastPlanAt: '2026-07-11T12:00:00Z',
  },
  {
    id: 'mock-f2',
    name: 'Sam Grower',
    email: 'sam.demo@farmsense.local',
    isActive: true,
    createdAt: '2026-03-02T10:00:00Z',
    lastLoginAt: null,
    district: 'Cumbria',
    region: 'Cumbria',
    farmSize: 2,
    countryCode: 'GB',
    hasSoilData: false,
    soilReadingsCount: 0,
    hasRecommendations: false,
    planStatus: 'none',
    finalized: false,
    topCrop: null,
    lastPlanAt: null,
  },
  {
    id: 'mock-f3',
    name: 'Jordan Fields',
    email: 'jordan.demo@farmsense.local',
    isActive: false,
    createdAt: '2026-02-14T10:00:00Z',
    lastLoginAt: '2026-04-01T09:00:00Z',
    district: 'North Yorkshire',
    region: 'North Yorkshire',
    farmSize: 8,
    countryCode: 'GB',
    hasSoilData: true,
    soilReadingsCount: 1,
    hasRecommendations: true,
    planStatus: 'draft',
    finalized: false,
    topCrop: 'Maize',
    lastPlanAt: '2026-06-20T12:00:00Z',
  },
  {
    id: 'mock-f4',
    name: 'Priya Mehta',
    email: 'priya.demo@farmsense.local',
    isActive: true,
    createdAt: '2026-01-22T11:30:00Z',
    lastLoginAt: '2026-07-10T07:15:00Z',
    district: 'County Durham',
    region: 'County Durham',
    farmSize: 6.2,
    countryCode: 'GB',
    hasSoilData: true,
    soilReadingsCount: 3,
    hasRecommendations: true,
    planStatus: 'finalized',
    finalized: true,
    topCrop: 'Potato',
    lastPlanAt: '2026-07-09T16:00:00Z',
  },
  {
    id: 'mock-f5',
    name: 'Chris Bancroft',
    email: 'chris.demo@farmsense.local',
    isActive: true,
    createdAt: '2025-11-05T09:00:00Z',
    lastLoginAt: '2026-07-08T18:40:00Z',
    district: 'Lancashire',
    region: 'Lancashire',
    farmSize: 12,
    countryCode: 'GB',
    hasSoilData: true,
    soilReadingsCount: 4,
    hasRecommendations: true,
    planStatus: 'finalized',
    finalized: true,
    topCrop: 'Wheat',
    lastPlanAt: '2026-07-07T10:20:00Z',
  },
  {
    id: 'mock-f6',
    name: 'Helen Crowe',
    email: 'helen.demo@farmsense.local',
    isActive: true,
    createdAt: '2026-04-18T14:00:00Z',
    lastLoginAt: '2026-07-01T12:00:00Z',
    district: 'Lincolnshire',
    region: 'Lincolnshire',
    farmSize: 18.5,
    countryCode: 'GB',
    hasSoilData: true,
    soilReadingsCount: 2,
    hasRecommendations: true,
    planStatus: 'draft',
    finalized: false,
    topCrop: 'Cabbage',
    lastPlanAt: '2026-06-28T09:45:00Z',
  },
  {
    id: 'mock-f7',
    name: 'Owen Hughes',
    email: 'owen.demo@farmsense.local',
    isActive: true,
    createdAt: '2026-02-01T08:20:00Z',
    lastLoginAt: '2026-06-30T11:10:00Z',
    district: 'Norfolk',
    region: 'Norfolk',
    farmSize: 9.1,
    countryCode: 'GB',
    hasSoilData: false,
    soilReadingsCount: 0,
    hasRecommendations: false,
    planStatus: 'none',
    finalized: false,
    topCrop: null,
    lastPlanAt: null,
  },
  {
    id: 'mock-f8',
    name: 'Megan Shaw',
    email: 'megan.demo@farmsense.local',
    isActive: true,
    createdAt: '2025-12-12T16:45:00Z',
    lastLoginAt: '2026-07-11T06:55:00Z',
    district: 'Cambridgeshire',
    region: 'Cambridgeshire',
    farmSize: 5.4,
    countryCode: 'GB',
    hasSoilData: true,
    soilReadingsCount: 1,
    hasRecommendations: true,
    planStatus: 'finalized',
    finalized: true,
    topCrop: 'Carrot',
    lastPlanAt: '2026-07-05T14:30:00Z',
  },
  {
    id: 'mock-f9',
    name: 'David Kerr',
    email: 'david.demo@farmsense.local',
    isActive: false,
    createdAt: '2026-03-28T10:00:00Z',
    lastLoginAt: '2026-05-02T09:00:00Z',
    district: 'East Riding of Yorkshire',
    region: 'East Riding of Yorkshire',
    farmSize: 14,
    countryCode: 'GB',
    hasSoilData: true,
    soilReadingsCount: 2,
    hasRecommendations: true,
    planStatus: 'draft',
    finalized: false,
    topCrop: 'Barley',
    lastPlanAt: '2026-05-01T13:00:00Z',
  },
  {
    id: 'mock-f10',
    name: 'Aisha Khan',
    email: 'aisha.demo@farmsense.local',
    isActive: true,
    createdAt: '2026-05-09T12:00:00Z',
    lastLoginAt: '2026-07-13T08:05:00Z',
    district: 'Kent',
    region: 'Kent',
    farmSize: 3.2,
    countryCode: 'GB',
    hasSoilData: true,
    soilReadingsCount: 1,
    hasRecommendations: true,
    planStatus: 'finalized',
    finalized: true,
    topCrop: 'Apple',
    lastPlanAt: '2026-07-12T17:00:00Z',
  },
  {
    id: 'mock-f11',
    name: 'Tom Reid',
    email: 'tom.demo@farmsense.local',
    isActive: true,
    createdAt: '2026-01-30T09:40:00Z',
    lastLoginAt: '2026-07-04T19:20:00Z',
    district: 'Devon',
    region: 'Devon',
    farmSize: 7.8,
    countryCode: 'GB',
    hasSoilData: false,
    soilReadingsCount: 0,
    hasRecommendations: false,
    planStatus: 'none',
    finalized: false,
    topCrop: null,
    lastPlanAt: null,
  },
  {
    id: 'mock-f12',
    name: 'Lucy Byrne',
    email: 'lucy.demo@farmsense.local',
    isActive: true,
    createdAt: '2026-04-02T15:10:00Z',
    lastLoginAt: '2026-07-06T10:00:00Z',
    district: 'Herefordshire',
    region: 'Herefordshire',
    farmSize: 11,
    countryCode: 'GB',
    hasSoilData: true,
    soilReadingsCount: 2,
    hasRecommendations: true,
    planStatus: 'draft',
    finalized: false,
    topCrop: 'Hop',
    lastPlanAt: '2026-07-03T11:25:00Z',
  },
  {
    id: 'mock-f13',
    name: 'Noah Patel',
    email: 'noah.demo@farmsense.local',
    isActive: true,
    createdAt: '2025-10-20T08:00:00Z',
    lastLoginAt: '2026-06-22T07:30:00Z',
    district: 'Shropshire',
    region: 'Shropshire',
    farmSize: 15.5,
    countryCode: 'GB',
    hasSoilData: true,
    soilReadingsCount: 5,
    hasRecommendations: true,
    planStatus: 'finalized',
    finalized: true,
    topCrop: 'Onion',
    lastPlanAt: '2026-06-21T15:40:00Z',
  },
  {
    id: 'mock-f14',
    name: 'Rachel Ng',
    email: 'rachel.demo@farmsense.local',
    isActive: true,
    createdAt: '2026-06-01T13:20:00Z',
    lastLoginAt: null,
    district: 'Greater London',
    region: 'Greater London',
    farmSize: 0.8,
    countryCode: 'GB',
    hasSoilData: false,
    soilReadingsCount: 0,
    hasRecommendations: false,
    planStatus: 'none',
    finalized: false,
    topCrop: null,
    lastPlanAt: null,
  },
  {
    id: 'mock-f15',
    name: 'Callum Forsyth',
    email: 'callum.demo@farmsense.local',
    isActive: true,
    createdAt: '2026-02-25T10:50:00Z',
    lastLoginAt: '2026-07-09T14:15:00Z',
    district: 'Scottish Borders',
    region: 'Scottish Borders',
    farmSize: 22,
    countryCode: 'GB',
    hasSoilData: true,
    soilReadingsCount: 2,
    hasRecommendations: true,
    planStatus: 'finalized',
    finalized: true,
    topCrop: 'Oats',
    lastPlanAt: '2026-07-08T09:00:00Z',
  },
]

function districtBreakdown(farmers) {
  const counts = {}
  for (const f of farmers) {
    const d = f.district || 'Unknown'
    counts[d] = (counts[d] || 0) + 1
  }
  return Object.entries(counts)
    .map(([district, count]) => ({ district, farmers: count }))
    .sort((a, b) => b.farmers - a.farmers || a.district.localeCompare(b.district))
}

function pct(part, whole) {
  if (!whole) return 0
  return Math.round((part / whole) * 1000) / 10
}

function dummyAnalytics() {
  const total = mockFarmers.length
  const active = mockFarmers.filter((f) => f.isActive).length
  const inactive = total - active
  const withSoil = mockFarmers.filter((f) => f.hasSoilData).length
  const finalized = mockFarmers.filter((f) => f.finalized || f.planStatus === 'finalized').length
  const draft = mockFarmers.filter((f) => f.planStatus === 'draft' && !f.finalized).length
  const noPlan = total - finalized - draft
  const byDistrict = districtBreakdown(mockFarmers)

  const cropCounts = {}
  for (const f of mockFarmers) {
    if ((f.finalized || f.planStatus === 'finalized') && f.topCrop) {
      cropCounts[f.topCrop] = (cropCounts[f.topCrop] || 0) + 1
    }
  }
  const topCrops = Object.entries(cropCounts)
    .map(([crop, farmers]) => ({ crop, farmers }))
    .sort((a, b) => b.farmers - a.farmers || a.crop.localeCompare(b.crop))

  const monthCounts = {}
  for (const f of mockFarmers) {
    if (!f.createdAt) continue
    const month = f.createdAt.slice(0, 7)
    monthCounts[month] = (monthCounts[month] || 0) + 1
  }
  const months = 'Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec'.split(' ')
  const signupsByMonth = Object.keys(monthCounts)
    .sort()
    .map((month) => {
      const [y, m] = month.split('-')
      return { month, label: `${months[Number(m) - 1]} ${y}`, farmers: monthCounts[month] }
    })

  const sizes = mockFarmers.map((f) => Number(f.farmSize)).filter((n) => Number.isFinite(n) && n > 0)
  const insights = []
  if (pct(withSoil, total) >= 60) {
    insights.push(`Soil data adoption is strong at ${pct(withSoil, total)}% of farmers.`)
  } else {
    insights.push(
      `Only ${pct(withSoil, total)}% of farmers have submitted soil data — worth nudging new sign-ups to complete Plan.`
    )
  }
  if (draft > finalized) {
    insights.push(
      `More draft plans (${draft}) than finalized (${finalized}) — farmers may need help confirming picks.`
    )
  } else {
    insights.push(`${pct(finalized, total)}% of farmers have a finalized crop plan.`)
  }
  if (inactive) insights.push(`${inactive} inactive account(s) are blocked from signing in.`)
  if (byDistrict[0]) {
    insights.push(`Most farmers are in ${byDistrict[0].district} (${byDistrict[0].farmers}).`)
  }
  if (topCrops[0]) {
    insights.push(`Leading finalized crop: ${topCrops[0].crop} (${topCrops[0].farmers} farm(s)).`)
  }

  return {
    mode: 'dummy',
    totals: {
      farmers: total,
      active,
      inactive,
      withSoilData: withSoil,
      draftPlans: draft,
      finalizedPlans: finalized,
      noPlan,
    },
    rates: {
      soilAdoptionPct: pct(withSoil, total),
      finalizedPct: pct(finalized, total),
      draftPct: pct(draft, total),
      noPlanPct: pct(noPlan, total),
      activePct: pct(active, total),
    },
    planStatus: [
      { status: 'Finalized', count: finalized },
      { status: 'Draft', count: draft },
      { status: 'No plan', count: noPlan },
    ],
    byDistrict,
    topCrops,
    signupsByMonth,
    farmSize: {
      avgHa: sizes.length ? Math.round((sizes.reduce((a, b) => a + b, 0) / sizes.length) * 100) / 100 : null,
      minHa: sizes.length ? Math.min(...sizes) : null,
      maxHa: sizes.length ? Math.max(...sizes) : null,
      reported: sizes.length,
    },
    insights: insights.slice(0, 5),
  }
}

function dummyOverview() {
  return {
    mode: 'dummy',
    totals: {
      farmers: mockFarmers.length,
      active: mockFarmers.filter((f) => f.isActive).length,
      inactive: mockFarmers.filter((f) => !f.isActive).length,
      withSoilData: mockFarmers.filter((f) => f.hasSoilData).length,
      draftPlans: mockFarmers.filter((f) => f.planStatus === 'draft').length,
      finalizedPlans: mockFarmers.filter((f) => f.finalized).length,
    },
    byDistrict: districtBreakdown(mockFarmers),
  }
}

export const adminService = {
  /**
   * @param {{ dataMode?: 'live' | 'dummy' }} [opts]
   */
  async getOverview({ dataMode = 'live' } = {}) {
    if (dataMode === 'dummy') {
      return withMockDelay(dummyOverview())
    }
    return backendClient.get(backendEndpoints.admin.overview)
  },

  async getAnalytics({ dataMode = 'live' } = {}) {
    if (dataMode === 'dummy') {
      return withMockDelay(dummyAnalytics())
    }
    return backendClient.get(backendEndpoints.admin.analytics)
  },

  /**
   * @param {{ q?: string, page?: number, limit?: number, dataMode?: 'live' | 'dummy' }} [opts]
   */
  async listFarmers({ q = '', page = 1, limit = 20, dataMode = 'live' } = {}) {
    if (dataMode === 'dummy') {
      const term = q.trim().toLowerCase()
      let items = mockFarmers
      if (term) {
        items = items.filter(
          (f) =>
            f.name.toLowerCase().includes(term) ||
            f.email.toLowerCase().includes(term) ||
            (f.district || '').toLowerCase().includes(term)
        )
      }
      const total = items.length
      const safeLimit = Math.max(1, Math.min(100, limit))
      const safePage = Math.max(1, page)
      const offset = (safePage - 1) * safeLimit
      const pageItems = items.slice(offset, offset + safeLimit)
      return withMockDelay({
        mode: 'dummy',
        items: pageItems,
        page: safePage,
        limit: safeLimit,
        total,
        totalPages: Math.max(1, Math.ceil(total / safeLimit)),
      })
    }
    return backendClient.get(backendEndpoints.admin.farmers, {
      q: q || undefined,
      page,
      limit,
    })
  },

  async getFarmer(id, { dataMode = 'live' } = {}) {
    if (dataMode === 'dummy') {
      const farmer = mockFarmers.find((f) => f.id === id) || mockFarmers[0]
      return withMockDelay({ ...farmer })
    }
    return backendClient.get(backendEndpoints.admin.farmer(id))
  },

  async setFarmerActive(id, isActive, { dataMode = 'live' } = {}) {
    if (dataMode === 'dummy') {
      const farmer = mockFarmers.find((f) => f.id === id) || mockFarmers[0]
      return withMockDelay({ ...farmer, isActive })
    }
    return backendClient.patch(backendEndpoints.admin.farmer(id), { isActive })
  },

  async deleteFarmer(id, { dataMode = 'live' } = {}) {
    if (dataMode === 'dummy') {
      return withMockDelay({ deleted: true, id, email: 'demo@farmsense.local', name: 'Demo' })
    }
    return backendClient.delete(backendEndpoints.admin.farmer(id))
  },

  async getModelsStatus() {
    return backendClient.get(backendEndpoints.admin.modelsStatus)
  },

  /**
   * @param {string} layer
   * @param {Record<string, unknown>} [payload]
   */
  async testModel(layer, payload = {}) {
    return backendClient.post(backendEndpoints.admin.modelsTest, { layer, payload })
  },
}
